webhook/mutation/pod_mutator.go

```diff
@@ -117,14 +124,81 @@ func registerHealthzEndpoint(mgr manager.Manager) {
 
 // podMutator injects the OneAgent into Pods
 type podMutator struct {
-	client    client.Client
-	apiReader client.Reader
-	decoder   *admission.Decoder
-	image     string
-	namespace string
-	apmExists bool
-	clusterID string
-	recorder  record.EventRecorder
+	client     client.Client
+	metaClient client.Client
+	apiReader  client.Reader
+	decoder    *admission.Decoder
+	image      string
+	namespace  string
+	apmExists  bool
+	clusterID  string
+	recorder   record.EventRecorder
+}
+
+func rootOwnerPod(ctx context.Context, cnt client.Client, pod *corev1.Pod, namespace string) (string, string, error) {
+	obj := &v1.PartialObjectMetadata{
+		TypeMeta: v1.TypeMeta{
+			APIVersion: pod.APIVersion,
+			Kind:       pod.Kind,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: pod.ObjectMeta.Name,
+			// pod.ObjectMeta.Namespace is empty yet
+			Namespace:       namespace,
+			OwnerReferences: pod.ObjectMeta.OwnerReferences,
+		},
+	}
+	return rootOwner(ctx, cnt, obj)
+}
+
+func rootOwner(ctx context.Context, cnt client.Client, o *v1.PartialObjectMetadata) (string, string, error) {
+	if len(o.ObjectMeta.OwnerReferences) == 0 {
+		kind := o.Kind
+		if kind == "Pod" {
+			kind = ""
+		}
+		return o.ObjectMeta.Name, kind, nil
+	}
+
+	om := o.ObjectMeta
+	for _, owner := range om.OwnerReferences {
+		if owner.Controller != nil && *owner.Controller && isWellKnownWorkload(owner) {
+			obj := &v1.PartialObjectMetadata{
+				TypeMeta: v1.TypeMeta{
+					APIVersion: owner.APIVersion,
+					Kind:       owner.Kind,
+				},
+			}
+			if err := cnt.Get(ctx, client.ObjectKey{Name: owner.Name, Namespace: om.Namespace}, obj); err != nil {
+				log.Error(err, "failed to query the object", "apiVersion", owner.APIVersion, "kind", owner.Kind, "name", owner.Name, "namespace", om.Namespace)
+				return o.ObjectMeta.Name, o.Kind, err
+			}
+
+			return rootOwner(ctx, cnt, obj)
+		}
+	}
+	return o.ObjectMeta.Name, o.Kind, nil
+}
+
+func isWellKnownWorkload(ownerRef v1.OwnerReference) bool {
+	knownWorkloads := []v1.TypeMeta{
+		{Kind: "ReplicaSet", APIVersion: "apps/v1"},
+		{Kind: "Deployment", APIVersion: "apps/v1"},
+		{Kind: "ReplicationController", APIVersion: "v1"},
+		{Kind: "StatefulSet", APIVersion: "apps/v1"},
+		{Kind: "DaemonSet", APIVersion: "apps/v1"},
+		{Kind: "Job", APIVersion: "batch/v1"},
+		{Kind: "CronJob", APIVersion: "batch/v1"},
+		{Kind: "DeploymentConfig", APIVersion: "apps.openshift.io/v1"},
+	}
+
+	for _, knownController := range knownWorkloads {
```

Comment:
since this is going to be called a lot, would putting the values into a map and just do a lookup help/be possible here?

Reply:
It seems that a map has lower efficiency than the list considering such a small number of items to be searched.